crates/guest-rust/rt/src/async_support/stream_support.rs

```diff
@@ -1,133 +1,109 @@
-extern crate std;
-
+use crate::async_support::waitable::{WaitableOp, WaitableOperation};
+use crate::async_support::AbiBuffer;
 use {
-    super::ErrorContext,
-    super::Handle,
-    futures::{
-        channel::oneshot,
-        future::{self, FutureExt},
-        sink::Sink,
-        stream::Stream,
-    },
+    crate::Cleanup,
     std::{
-        boxed::Box,
-        collections::hash_map::Entry,
-        convert::Infallible,
+        alloc::Layout,
         fmt,
         future::Future,
-        iter,
-        mem::{self, MaybeUninit},
+        marker,
         pin::Pin,
+        ptr,
         sync::atomic::{AtomicU32, Ordering::Relaxed},
         task::{Context, Poll},
         vec::Vec,
     },
 };
 
-fn ceiling(x: usize, y: usize) -> usize {
-    (x / y) + if x % y == 0 { 0 } else { 1 }
-}
-
 #[doc(hidden)]
 pub struct StreamVtable<T> {
-    pub write: fn(future: u32, values: &[T]) -> Pin<Box<dyn Future<Output = usize> + '_>>,
-    pub read: fn(
-        future: u32,
-        values: &mut [MaybeUninit<T>],
-    ) -> Pin<Box<dyn Future<Output = Option<Result<usize, ErrorContext>>> + '_>>,
-    pub cancel_write: unsafe extern "C" fn(future: u32) -> u32,
-    pub cancel_read: unsafe extern "C" fn(future: u32) -> u32,
-    pub close_writable: unsafe extern "C" fn(future: u32, err_ctx: u32),
-    pub close_readable: unsafe extern "C" fn(future: u32, err_ctx: u32),
-    pub new: unsafe extern "C" fn() -> u32,
+    pub layout: Layout,
+    pub lower: Option<unsafe fn(value: T, dst: *mut u8)>,
+    pub dealloc_lists: Option<unsafe fn(dst: *mut u8)>,
+    pub lift: Option<unsafe fn(dst: *mut u8) -> T>,
```

Comment:
I'll note that these are crucially `Option` which indicates that the raw representation of `T` can be used to avoid lifting/lowering overhead, aka the `stream<u8>` optimization. I should document this.